app/src/org/commcare/connect/network/connectId/parser/RetrieveNotificationsResponseParser.kt

```diff
@@ -1,19 +1,111 @@
+                if (channel.getConsented() && channel.getKey().length == 0) {
+                    needReloadDueToMissingChannelKey = true
+                    excludeMessagesForChannel(channel.channelId)
+                    MessageManager.getChannelEncryptionKey(context, channel, null)
+                }
+            }
+        }
+
+        notificationsJsonArray?.let {
+            val existingChannels = ConnectMessagingDatabaseHelper.getMessagingChannels(context)
+            for (notificationJsonIndex in 0 until notificationsJsonArray!!.length()) {
+                val notificationJsonObject =
+                    notificationsJsonArray!!.getJSONObject(notificationJsonIndex)
+                if (pushNotificationIsMessageType(notificationJsonObject) &&
+                    notificationJsonObject.getJSONObject("data").has("message_id")
+                ) {
+                    val message = ConnectMessagingMessageRecord.fromJson(
+                        notificationJsonObject.getJSONObject("data"),
+                        existingChannels
+                    )
+                    if (message != null) {
+                        messages.add(message)
+                    }
+                }
+            }
+            ConnectMessagingDatabaseHelper.storeMessagingMessages(context, messages, false)
+        }
+
+        if (needReloadDueToMissingChannelKey) {
+            Handler(Looper.getMainLooper()).postDelayed({
+                scheduleImmediatePushNotificationRetrieval(CommCareApplication.instance())
+            }, 3000)
+        }
+
+        val result = PushNotificationRecord.fromJsonArray(notificationsJsonArray ?: JSONArray())
         return result as T
     }
+
+    private fun excludeMessagesForChannel(channelId: String) {
+        val newNotificationsJsonArray = JSONArray()
+        notificationsJsonArray?.let {
+            for (notificationJsonIndex in 0 until notificationsJsonArray!!.length()) {
+                var removeThisObject = false
+                val notificationJsonObject =
+                    notificationsJsonArray!!.getJSONObject(notificationJsonIndex)
+                if (pushNotificationIsMessageType(notificationJsonObject) &&
+                    notificationJsonObject.getJSONObject("data").get("channel") != null &&
+                    notificationJsonObject.getJSONObject("data").get("channel").equals(channelId)
+                ) {
+                    removeThisObject = true
+                }
+                if (!removeThisObject) {
+                    newNotificationsJsonArray.put(notificationJsonObject)
+                }
+            }
+        }
+        notificationsJsonArray = newNotificationsJsonArray
+    }
+
+    private fun pushNotificationIsMessageType(notificationJsonObject: JSONObject) =
+        notificationJsonObject.has("notification_type") && "MESSAGING".equals(
```

Comment:
did we devide to have a `notification_type` in payload ? My sense was we are trying to figure this with notification `action`